### PR TITLE
Add MaGe dario .root.hits reader

### PR DESCRIPTION
**Pull request by Luapulu**
Adds support for .root.hits files. Tests are comprehensive and easily reach 100% test coverage.

### CODE REVIEW

**Comment by Luapulu** [comment]
Should be good to merge now

**Comment by Luapulu** [comment]
Note that right now the code doesn't work on windows. Works on linux and mac os though.

**Comment by oschulz** [comment]
> Note that right now the code doesn't work on windows.

Why not? In principle, we would like to cover all three platforms as far as possible, of course.

**Comment by Luapulu** [comment]
Honestly, because I have a mac and just haven't gotten around to setting up a virtual machine so I can debug the issue.

**Comment by oschulz** [comment]
Ah, right - well, we can tackle that when we have CI tests on GitHub.

File: src/LegendTextIO.jl [hunk 1]
```diff
@@ -19,4 +19,15 @@ using RadiationDetectorSignals: group_by_evtno
 include("util.jl")
 include("geant4_csv.jl")
 
+## .root.hit files
```
Comment (oschulz) [inline src/LegendTextIO.jl]: Please use lower case for source file names, so "root_hist_file.jl" or so in this case.

**Comment by Luapulu** [comment]
I'm gonna setup a virtual machine now. So I think I can probably fix the issue today.

File: src/RootHitFiles.jl [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    RootHitFile(file::Union{IOStream, AbstractString})
+
+represents a `.root.hits` file, which can be iterated or read to yield events, containg data
+on hits.
+"""
+struct RootHitFile
+    stream::IOBuffer
+end
+
+RootHitFile(stream::IOStream) = RootHitFile(IOBuffer(mmap(stream)))
```
Comment (oschulz) [inline src/RootHitFiles.jl]: Does mmap actually improve performance here?

File: src/RootHitFiles.jl [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    eof(f) && return nothing
+    return read(f), nothing
+end
+
+Base.IteratorSize(::Type{RootHitFile}) = Base.SizeUnknown()
+Base.IteratorEltype(::Type{RootHitFile}) = Base.HasEltype()
+Base.eltype(::RootHitFile) = EventTuple
+
+Tables.isrowtable(::Type{RootHitFile}) = true
+Tables.schema(f::RootHitFile) = Tables.Schema(eltype(f))
```
Comment (oschulz) [inline src/RootHitFiles.jl]: Interesting approach, defining an on-disk table. Not sure about how far it will carry us here, because the file format (in contrast to HDF5) doesn't allow for random access. But let's try it. :-)

File: test/runtests.jl [hunk 4]
```diff
@@ -1,8 +1,9 @@
 # This file is a part of LegendTextIO.jl, licensed under the MIT License (MIT).
 
-import Test
-Test.@testset "Package LegendTextIO" begin
+using Test, LegendTextIO, Tables
 
-# test code goes here
+@testset "LegendTextIO.jl" begin
```
Comment (oschulz) [inline test/runtests.jl]: Please use `Test.@testset "Package LegendTextIO" begin`

File: test/test_root_hit.jl [hunk 5]
```diff
@@ -0,0 +1,115 @@
+@testset "RootHitFiles" begin
```
Comment (oschulz) [inline test/test_root_hit.jl]: Should be `"@testset "RootHitFile" begin`

File: test/test.root.hits [hunk 6]
```diff
@@ -0,0 +1,1181 @@
+624 119 3
```
Comment (oschulz) [inline test/test.root.hits]: I guess this file still has to be taken out (now in legend-testdata), right?

**Review by oschulz — CHANGES_REQUESTED** [review]
Nice - I marked a few things that should still be changed.

File: src/RootHitFiles.jl (hunk 2 above)
Comment (Luapulu) [reply src/RootHitFiles.jl]: I tried a bunch of different things and memory mapped IOBuffer was the best (by quite a lot) when I last ran benchmarks. But now that you ask I'm curious again, so I might see if that's changed.

File: src/RootHitFiles.jl (hunk 3 above)
Comment (Luapulu) [reply src/RootHitFiles.jl]: We actually need out-of-memory capability because it allows us to run a lot more processes in parallel, thus significantly improving performance.

File: test/runtests.jl (hunk 4 above)
Comment (Luapulu) [reply test/runtests.jl]: This is ultimately a taste thing, so sure. I am curious though. Why? I think my version is the more common style and simpler, no?

Do you want `Test.@testset` for every testset by the way?

File: test/test_root_hit.jl (hunk 5 above)
Comment (Luapulu) [reply test/test_root_hit.jl]: Sure. This is a vestige of it being its own package

File: test/test.root.hits (hunk 6 above)
Comment (Luapulu) [reply test/test.root.hits]: Yeah this file can be moved elsewhere if you don't want the test data here.

File: src/LegendTextIO.jl (hunk 1 above)
Comment (Luapulu) [reply src/LegendTextIO.jl]: Will do.

File: src/RootHitFiles.jl (hunk 3 above)
Comment (Luapulu) [reply src/RootHitFiles.jl]: And of course you can always do `collect(RootHitFile("path"))`, giving you the same thing. :)

File: src/RootHitFiles.jl (hunk 2 above)
Comment (Luapulu) [reply src/RootHitFiles.jl]: Ok here are the comparisons for using a straight up `IOStream` vs a `IOBuffer` of a memory mapped file.

## TLDR

Time to collect the entire file goes from 6.254 ms without mmap to 0.2672 ms with mmap.

Time to read first event goes from 644.408 μs without mmap to 41.231 μs.

Time improvement comes at the cost of about 18% more total memory allocation.

## Benchmarks

I used the following three benchmarks:

```julia
SUITE = BenchmarkGroup()

SUITE["construct"] = @benchmarkable RootHitFile($eventpath)

SUITE["read"] = @benchmarkable read(RootHitFile($eventpath))

SUITE["collect"] = @benchmarkable collect(RootHitFile($eventpath))
```

### IOStream (no memory mapping)

```julia
julia> r = run(SUITE)
3-element BenchmarkTools.BenchmarkGroup:
  tags: []
  "construct" => Trial(9.499 μs)
  "collect" => Trial(6.254 ms)
  "read" => Trial(644.408 μs)

julia> r["collect"]
BenchmarkTools.Trial: 
  memory estimate:  203.45 KiB
  allocs estimate:  2444
  --------------
  minimum time:     6.254 ms (0.00% GC)
  median time:      6.743 ms (0.00% GC)
  mean time:        6.806 ms (0.25% GC)
  maximum time:     27.951 ms (0.00% GC)
  --------------
  samples:          735
  evals/sample:     1

julia> r["read"]
BenchmarkTools.Trial: 
  memory estimate:  21.48 KiB
  allocs estimate:  261
  --------------
  minimum time:     644.408 μs (0.00% GC)
  median time:      683.241 μs (0.00% GC)
  mean time:        698.904 μs (0.34% GC)
  maximum time:     15.795 ms (37.35% GC)
  --------------
  samples:          7145
  evals/sample:     1
```

### IOBuffer (memory mapped)

```julia
julia> r = run(SUITE)
3-element BenchmarkTools.BenchmarkGroup:
  tags: []
  "construct" => Trial(14.125 μs)
  "collect" => Trial(267.160 μs)
  "read" => Trial(41.231 μs)

julia> r["collect"]
BenchmarkTools.Trial: 
  memory estimate:  239.48 KiB
  allocs estimate:  2435
  --------------
  minimum time:     267.160 μs (0.00% GC)
  median time:      274.981 μs (0.00% GC)
  mean time:        302.123 μs (3.58% GC)
  maximum time:     19.460 ms (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1

julia> r["read"]
BenchmarkTools.Trial: 
  memory estimate:  25.64 KiB
  allocs estimate:  266
  --------------
  minimum time:     41.231 μs (0.00% GC)
  median time:      43.882 μs (0.00% GC)
  mean time:        52.881 μs (3.83% GC)
  maximum time:     15.698 ms (27.84% GC)
  --------------
  samples:          10000
  evals/sample:     1
```

**Comment by Luapulu** [comment]
@oschulz I'll be curious what you think of the benchmarks.

I made the changes you asked for. I went with hyphens over underscores, because file names allow them and I think they look a lot better.

Oh, before I forget. I know the code works with julia 1.3-1. I don't remember if it works with 1.0.

**Comment by Luapulu** [comment]
Btw, do you want me to add the benchmarks, since I have them in the original package?

File: src/RootHitFiles.jl (hunk 2 above)
Comment (oschulz) [reply src/RootHitFiles.jl]: Could you include BufferedStreams.jl in the comparison (with and without mmap)? That could be valuable for similar ASCII-handling cases.

Comment (oschulz) [reply src/RootHitFiles.jl]: I've been using it here already: https://github.com/legend-exp/LegendTextIO.jl/blob/268cb42a1cf6f80ca429d74fc0bf11ca72adc515/src/geant4_csv.jl#L14

File: test/runtests.jl (hunk 4 above)
Comment (oschulz) [reply test/runtests.jl]: >  I am curious though. Why?

Comment (oschulz) [reply test/runtests.jl]: > I am curious though. Why?

Just because I/we've been using it usually so far. But you're right, it doesn't matter much.

> Do you want Test.@testset for every testset by the way

Every file included by "runtests.jl" should contain a testset with the name of the file (without "test_" and ".jl" in the title). Inside those, use tests files to structure things, but not excessively.

File: test/test.root.hits (hunk 6 above)
Comment (oschulz) [reply test/test.root.hits]: Yes, please - we just did merge it into legend-testdata, after all. :-) I'll see that legend-testdata and LegendTestData.jl get a proper master branch over the weekend.

File: src/RootHitFiles.jl (hunk 3 above)
Comment (oschulz) [reply src/RootHitFiles.jl]: > We actually need out-of-memory capability

Yes, of course. I was just wondering if the Tables API is the right choice for streaming input, but let's give it a go and find out in practice. Could you also add support for `Tables.partitions`, though, to allow block-wise processing? The blocks should be returned as `TypedTabes.Table` or `StructArrays.StructArray`. Block-size could be an option set by the user when opening the file. I have similar plans for our HDF5-I/O.

File: src/root-hits.jl [hunk 7]
```diff
@@ -0,0 +1,71 @@
+"""
```
Comment (oschulz) [inline src/root-hits.jl]: Please do use underscores, not hyphens, it's way more common to uses underscores in the Julia community.

**Comment by oschulz** [comment]
> Oh, before I forget. I know the code works with julia 1.3-1. I don't remember if it works with 1.0.

**Comment by oschulz** [comment]
> Oh, before I forget. I know the code works with julia 1.3-1. I don't remember if it works with 1.0.

Just set Julia v1.3 as the minumum Julia version. We'll likely start to require v1.6 as minimum as soon as v1.7 is out anyhow (v1.6 is likely to become the new LTS Julia version at that point).

**Comment by oschulz** [comment]
> Btw, do you want me to add the benchmarks, since I have them in the original package?

No, let's keep it simple.

File: src/RootHitFiles.jl (hunk 3 above)
Comment (Luapulu) [reply src/RootHitFiles.jl]: I can add `Tables.partitions`, a global default batch size the user can change (like BenchmarkTools.jl does it) and a kw argument to the file to set it manually. That way it's just as easy to use and works similar to functions like pmap where it's also a kw argument.

I'm against returning any particular table though. I'd prefer returning an iterable of `Tables.rowtable`, meaning an iterable of vectors of `NamedTuples`. It's the simplest most straightforward solution, which doesn't add any further dependencies. It also costs nothing for the user to convert to their preferred table type vs. me doing it for them.

File: test/test.root.hits (hunk 6 above)
Comment (Luapulu) [reply test/test.root.hits]: I think the one in legend testdata might actually be different to this one if you merged my earlier PR. I'll check and make a new one if needed.

File: src/root-hits.jl (hunk 7 above)
Comment (Luapulu) [reply src/root-hits.jl]: Ok, I actually just checked some packages and base, because I was sceptical. The *most* common trend seems to be to have short file names of one word or join two words together like "fastmath.jl" in julia base. Almost all names follow that convention. You're right that, when two words cannot be joined in a readable way, they are separated by underscores.

Given this I'd name the main script: "RootHitFile.jl" and the test script "test_RootHitFile.jl", since files are usually named after the central type or function of that script (if there is a central type or function).

File: src/RootHitFiles.jl (hunk 2 above)
Comment (Luapulu) [reply src/RootHitFiles.jl]: Will do

File: src/LegendTextIO.jl [hunk 8]
```diff
@@ -19,4 +19,21 @@ using RadiationDetectorSignals: group_by_evtno
...
+
+export RootHitFile
+
+include("RootHitFile.jl")
+
+mutable struct Defaults
+	root_hits_batch_size::Int
+end
+
+const DEFAULTS = Defaults(10)
```
Comment (Luapulu) [inline src/LegendTextIO.jl]: I really like this way of defining defaults. It allows the user to change them if they want and it's easy to expand.

File: test/test_RootHitFile.jl [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+        p3 = collect(Iterators.partition(RootHitFile("test.root.hits"), 3))
+
+        @test length(p1) == length(p2) == length(p3)
+        @test eltype(p1) == eltype(p2) == eltype(p3)
+
+        @test eltype(p1) <: Vector{<:NamedTuple}
+
+        for (tbl1, tbl2, tbl3) in zip(p1, p2, p3)
+            @test tbl1 == tbl2 == tbl3
+            @test all(isa.((tbl1, tbl2, tbl3), Vector{<:NamedTuple}))
```
Comment (Luapulu) [inline test/test_RootHitFile.jl]: I might get rid of this test. Presumably we can trust that a collected vector will actually have the element type `eltype` specifies

File: src/RootHitFiles.jl (hunk 3 above)
Comment (oschulz) [reply src/RootHitFiles.jl]: > global default batch size the user can change (like BenchmarkTools.jl does it) and a kw argument to the file to set it manually

Comment (oschulz) [reply src/RootHitFiles.jl]: > global default batch size the user can change (like BenchmarkTools.jl does it) and a kw argument to the file to set it manually

A kwarg with a default should to nicely.

> I can add Tables.partitions [...] I'm against returning any particular table though.

Here, we do want to introduce certain standards across the LEGEND Julia ecosystem, to provide consistency to the users. It's perfectly fine if `RootHitFile` (you do, in fact, define a "particular table" here :-) ) is "just" an iterable over `EventTuple` (needs to be renamed, too generic) that implements the `Table` interface. However, `Tables.partitions` should return a an iterable over random access tables, and those should have a column-based memory layout. `StructArrays.StructArray` is the ideal choice here (and there's no alternative package) if you want structs, and `TypedTables.Table` if you want `NamedTuple`s.

File: src/RootHitFile.jl [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
+function RootHitFile(path::AbstractString; batch_size=DEFAULTS.root_hits_batch_size)
+    if occursin(r".root.hits$", path)
+        return RootHitFile(open(path); batch_size=batch_size)
+    else
+        throw(ArgumentError("$path is not a .root.hits file"))
+    end
+end
+
+EventTuple = NamedTuple{
```
Comment (oschulz) [inline src/RootHitFile.jl]: `EventTuple` is way too generic here, other people may have tuples over different kinds of events. :-)

Also, ideally use `const` when declaring type aliases. One thing to consider: Type aliases for `NamedTuple` can only used in limited ways in dispatch without committing type piracy.

**Review by oschulz — CHANGES_REQUESTED** [review]

File: test/test.root.hits (hunk 6 above)
Comment (oschulz) [reply test/test.root.hits]: Sure, whichever file you think best, as long as it isn't too big.

File: src/root-hits.jl (hunk 7 above)
Comment (oschulz) [reply src/root-hits.jl]: Well, *my* convention is that we use lower case file names, use underscores where a file relates to a struct with camel-case, and use "test_..." for test code (one test code file per source file) to that if you have multiple files open in editor tabs you can quickly see what's package and what's test code. And I'd like to to have some consistency across the legend packages. :-)

File: src/RootHitFile.jl [hunk 11]
```diff
@@ -1,16 +1,29 @@
 """
-    RootHitFile(file::Union{IOStream, AbstractString})
+    RootHitFile(file::Union{IOStream, AbstractString}; batch_size=10)
```
Comment (oschulz) [inline src/RootHitFile.jl]: Better to use `base_size::Integer = 10`

File: src/LegendTextIO.jl (hunk 8 above)
Comment (oschulz) [reply src/LegendTextIO.jl]: It's ok to use a global variable for the default batch size, but please do it like this:

```julia
const DEFAULT_BATCH_SIZE = Ref(10)
```

File: src/RootHitFile.jl (hunk 10 above)
Comment (oschulz) [reply src/RootHitFile.jl]: If you don't want to dispatch on it, using a `NamedTuple` is perfectly fine, of course. But if you define it this tightly, you probably actually want to define a struct instead.

File: src/RootHitFile.jl [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+end
+
+Base.IteratorSize(::Type{RootHitFile}) = Base.SizeUnknown()
+Base.IteratorEltype(::Type{RootHitFile}) = Base.HasEltype()
+Base.eltype(::Type{RootHitFile}) = EventTuple
+
+Tables.isrowtable(::Type{RootHitFile}) = true
+Tables.schema(f::RootHitFile) = Tables.Schema(eltype(f))
+
+Tables.partitions(f::RootHitFile) = Iterators.partition(f, f.batch_size)
```
Comment (oschulz) [inline src/RootHitFile.jl]: Please to return an iterator over `TypedTables.Table` or `StructArrays.StructArray` here, depending on whether the (renamed) `EventTuple` will be a `NamedTuple` or `struct`.

**Review by oschulz — CHANGES_REQUESTED** [review]

File: src/LegendTextIO.jl (hunk 8 above)
Comment (Luapulu) [reply src/LegendTextIO.jl]: Sure. Will do. :)

File: src/RootHitFile.jl (hunk 10 above)
Comment (Luapulu) [reply src/RootHitFile.jl]: Oh yeah, forgot to make it a const. Thanks :)

And I'll rename it to `RootHitEventTuple` or something like that so as to make clear it's specific to this file format.

And yeah, it's a `NamedTuple` exactly because nothing dispatches on the type.

File: src/RootHitFile.jl (hunk 11 above)
Comment (Luapulu) [reply src/RootHitFile.jl]: Sure, I can add the type annotation. :)

File: src/root-hits.jl (hunk 7 above)
Comment (Luapulu) [reply src/root-hits.jl]: Ah ok. So "root_hits.jl" and "test_root_hits.jl" for example would be your preference?

File: src/RootHitFiles.jl (hunk 3 above)
Comment (Luapulu) [reply src/RootHitFiles.jl]: TLDR: I'm happy to do what you suggest. I was missing something in my thinking earlier.

Here's what I was missing: I can write the partitions into a column based format directly instead of writing a row table and then converting. I was always thinking of writing my tuples into a vector and then transforming it to a column-based format, which would be inefficient compared to just leaving it in the original partitioned format. That's why I was sceptical.

I'll have to add a little convenience method to allocate a NamedTuple of vectors of a given length. I guess this is why you suggested StructArrays, because they have this built in?

File: src/root-hits.jl (hunk 7 above)
Comment (oschulz) [reply src/root-hits.jl]: Yes, resp. "root_hist_file.jl" if the main type in the file is called RootHistFile.

Question, though - is that name really ideal? People will assume it's about ".root" files that contains hits. Shouldn't we have something with "MaGe" and "Dario" in the name of the type instead?

File: src/RootHitFiles.jl (hunk 3 above)
Comment (oschulz) [reply src/RootHitFiles.jl]: > I was always thinking of writing my tuples into a vector and then transforming it to a column-based format, which would be inefficient

Comment (oschulz) [reply src/RootHitFiles.jl]: > writing my tuples into a vector and then transforming it to a column-based format, which would be inefficient

Indeed, you'll want to build column-based storage directly for each partition, of course. However, in practice, the fact that we parse a non-column-based text file as input will probably be what limits the performance. :-)

File: src/root_hits.jl [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+        throw(ArgumentError("$path is not a .root.hits file"))
+    end
+end
+
+const ROOT_HITS_BATCH_SIZE = Ref(10)
+
+const RootHitEventTuple = NamedTuple{
+    (:eventnum, :primcount, :pos, :E, :time, :particleID, :trkID, :trkparentID, :volumeID),
+    Tuple{
+        Int32, Int32, Vector{NTuple{3, Float32}}, Vector{Float32}, Vector{Float32},
```
Comment (oschulz) [inline src/root_hits.jl]: Might be good to use a `StaticArrays.SVector{3}` instead of an `NTuple{3}`, since `SVector` supports LinearAlgebra (we'll want to translate, rotate, etc. hit positions).

File: src/RootHitFiles.jl (hunk 3 above)
Comment (oschulz) [reply src/RootHitFiles.jl]: > I'll have to add a little convenience method to allocate a NamedTuple of vectors of a given length. I guess this is why you suggested StructArrays, because they have this built in?

No, StructArrays.jl is about storing arrays of generic strucs SoA-fashion, and TypesTables.jl is about storing arrays of `NamedTuple`s SoA-fashion (column-wise).

If you create an empty `TypedTables.Table` you can push into it. If you do know the number of rows, you can also pre-allocate the storage, of course.

To be efficient, though, you should use an `ArraysOfArrrays.VectorOfVectors` to back all columns that contain vectors. Otherwise you're be allocation a separate vector for each vector-valued field in `RootHitEventTuple`, for each row. Since those vectors will usually be short, that would be very wasteful. `VectorOfVectors` stores a vector of variable-length vectors as a single contiguous vector under the hood. I use it for waveforms, too.

Btw: Longer term, we need to look at all the MC-inputs we have now and standardize column names and types (the ones they have in common, semantically).

File: src/root-hits.jl (hunk 7 above)
Comment (Luapulu) [reply src/root-hits.jl]: Yeah, good point, though I'm having trouble coming up with an alternative that's clear and concise. The fact that it's in the TextIO package should help make it clear that it's not a true .root format. I guess the full description would be: `MaGeDarioRootHitFile`, but that's pretty ugly. You could go down to `DarioRootHitFile`, but that's kinda long, too. Another option would be to encase that script in a module called `Dario`, so you'd have `Dario.RootHitFile` ?

File: src/RootHitFiles.jl (hunk 3 above)
Comment (Luapulu) [reply src/RootHitFiles.jl]: Oh yeah, `VectorOfVectors` is a good idea.

And yes, I was going to ask about the naming. My naming of columns is only partly based on what I saw legend packages using elsewhere, but we should figure out naming conventions for this stuff.

> If you create an empty TypedTables.Table you can push into it. If you do know the number of rows, you can also pre- allocate the storage, of course.

What I meant is that StructArrays supports `StructArray{MyStruct}(undef, 5)`. Does TypedTables have something like that? I didn't find it in the source. Either way, I'll have to make a custom method to be able to use `VectorOfVectors`.

Also, for the VolID field I might go to something like a categorical array, to avoid having a `Vector{String}`.

On the performance question, it's true that converting everything to a column based format does slow things down a bit. The main bottleneck is still parsing of text though. The old version of the code was fully row based, so I have the old benchmarks somewhere.

File: src/root_hits.jl (hunk 13 above)
Comment (Luapulu) [reply src/root_hits.jl]: I had this before. I think I had a phase where I wanted to simplify everything so I threw it out. Probably went too far on my quest for simplification on that point.

This, by the way, is why I don't have a `:32` / `:64` type parameter to specify parsing as 32 bit or 64 bit ints / floats. It's simpler and until I'm convinced anybody actually needs 64 bit integers / precision I might as well save myself the trouble of implementing it. Would be curious if you think it's worth adding the extra complexity.

File: src/root-hits.jl (hunk 7 above)
Comment (oschulz) [reply src/root-hits.jl]: I think a submodule would be overkill here. How about `MaGeDarioOutputFile` or so?

File: src/RootHitFiles.jl (hunk 3 above)
Comment (oschulz) [reply src/RootHitFiles.jl]: > And yes, I was going to ask about the naming. My naming of columns

Just choose something for now. We'll iterate with @sagitta42 afterwards to harmonize things across input formats.

Comment (oschulz) [reply src/RootHitFiles.jl]: > Does TypedTables have something like that?

Comment (oschulz) [reply src/RootHitFiles.jl]: > What I meant is that StructArrays supports StructArray{MyStruct}(undef, 5). Does TypedTables have something like that?

No, you have to create the column vectors manually. But you'll have to do that anyhow to ensure you use column types like `VectorOfVectors` (it's semantic knowledge, neither TypedTables not StructArrays can guess that).

Comment (oschulz) [reply src/RootHitFiles.jl]: > Also, for the VolID field I might go to something like a categorical array, to avoid having a Vector{String}

Yes, `String`s are out of the question. Is it a finite small set of values? Then we can use an enum. Otherwise we'll have to use a caterorical array (I think there are two or three different ones on the Julia ecosystem currently).

Comment (oschulz) [reply src/RootHitFiles.jl]: > it's true that converting everything to a column based format does slow things down a bit. 

No matter, since at one point we'd want to go to columns anyhow.

File: src/root_hits.jl (hunk 13 above)
Comment (oschulz) [reply src/root_hits.jl]: I agree, in a use case like this we should just choose a precision and stick with it. It's different for computational code that may have to handle dual numbers and things like this, but for input, fixed types are usually fine.

File: src/root-hits.jl (hunk 7 above)
Comment (Luapulu) [reply src/root-hits.jl]: It's not the only output file from Dario. You also get actual ".root" files as part of the output.

File: src/RootHitFile.jl (hunk 12 above)
Comment (Luapulu) [reply src/RootHitFile.jl]: Ok, I started working on this a bit. This would require quite a bit of additional code for a feature that the main user of this piece of code (me) probably won't use. I'd suggest leaving this out of this PR for now. Of course, if people start using this and want this feature, I'd be happy to help implement this or give guidance, but right now I don't see this having adequate return on investment.

Let me know if you'd prefer to keep the partition operator the way it's now or if you want to just get rid of it and leave it open to implement later.

File: src/root-hits.jl (hunk 7 above)
Comment (oschulz) [reply src/root-hits.jl]: Sure, but we're not reading those here, that will go into LegendUpROOTIO.jl :-)

File: src/RootHitFile.jl (hunk 12 above)
Comment (oschulz) [reply src/RootHitFile.jl]: Why would it require so much extra code to include the conversion to a `TypedTables.Table` in the partitions iterator? Since we're going with `NamedTuple`s, we won't use `StructArray`s here.

Comment (Luapulu) [reply src/RootHitFile.jl]: As I see it I'd need / want to change three aspects of the code, for the purpose of making efficient column based partitioned iteration work.

1. To make this efficient I can't use the `Iterators.partition` because it is an eager iterator and returns vectors of `eltype` as partitions. I want to write directly into a TypedTable, rather than a vector of named tuples (and then converting to TypedTable).
2. In order to not return a vector of named tuples (and convert to TypedTable after), I need to write my own partition iterator class (something like `PartionedRootHitFile`) where I define how it should allocate an empty table and then write into it on each iteration of the underlying file until the partition is full or the file has ended.
3. To make writing into a VectorOfVector work efficiently I have to add a mutating `read!` function which will read into some kind of NamedTuple. Then I can use the EventTuple I defined in the normal case for `read` and a special tuple, using VectorOfVectors for the partitioned case. If I didn't do this it would write into a `Vector` and then have to copy the vector into the relevant part of the `VectorOfVector`.

Assuming I didn't miss some obvious way to do this in a much simpler way, I'd say that these changes should be moved to a separate PR someday. All of these are reasonable additions to make, but it's a significant bit of code to write and test. If you like we can open an issue for it now and lay out this plan so that if, in future, people come across this and want the feature they can see that this was discussed.

Note, if you want I can add a mutating `read!`, since that's a nice thing to have regardless.

File: src/root-hits.jl (hunk 7 above)
Comment (Luapulu) [reply src/root-hits.jl]: I think it's a toss up if `MaGeDarioOutputFile` or `RootHitFile` is clearer. So, on the basis that `RootHitFile` is shorter and looks nicer I'd pick that.

**Comment by Luapulu** [comment]
At this point I'd be in favour of merging (I'd like to switch to using this rather than my old package).

For the near future I'd want to:

- switch `Vector{String}` with a categorical array. I'm not sure how limited the options are for the volume, and what they even are, so I'd stick with just categorical array for now.
- check if `BufferedInputStream` is faster. 
- Add a `read!` method
- Make PR to legend testdata for the testdata here

For the intermediate future:

- Fix bug on windows
- Implement efficient partioned column table iteration
- Switch to using legend test data rather than having data here

**Comment by Luapulu** [comment]
Another three points to work on at some point:

- column naming conventions
- documenting what the fields in the event mean
- documenting some basic usage example

File: src/RootHitFile.jl (hunk 12 above)
Comment (oschulz) [reply src/RootHitFile.jl]: Well, this actually doesn't have to be super performant, because the parsing will still be the step that limits speed. But sure, we can do this later on.

File: src/root-hits.jl (hunk 7 above)
Comment (oschulz) [reply src/root-hits.jl]: Sorry, but I really don't think people will even know it's from MaGe if called `RootHitFile`. The name should be intuitive to new users. It is MaGe's dario format right? Which specific to MaGe, not a Geant4-thing in general?

**Comment by oschulz** [comment]
> At this point I'd be in favour of merging (I'd like to switch to using this rather than my old package). [...]

Sure, all of the other stuff can be done later. However, the name of the class we should change before merging.

File: src/root-hits.jl (hunk 7 above)
Comment (Luapulu) [reply src/root-hits.jl]: Ah, I have an idea. How about `DarioHitFile` ? Since it's the file type that Dario outputs that gives you information on individual hits as opposed to summary info on whole events. That name would tell you what it contains and where it comes from. In the docstring you'd say it's from MaGe and has file ending ".root.hits".

And I think you're right about it being specific to MaGe as opposed to being a general geant4 format, but I don't know for certain.

**Comment by Luapulu** [comment]
Ok, I made a naming suggestion above. Let me know what you think :)

File: src/root-hits.jl (hunk 7 above)
Comment (Luapulu) [reply src/root-hits.jl]: Or the plural: `DarioHitsFile`?

Comment (Luapulu) [reply src/root-hits.jl]: I should add that the other files Dario outputs only give info on the event as a whole, like R90 or the event number. So, the root.hits format is truly the "hits" data format of the Dario output.

**Comment by Luapulu** [comment]
I made a naming suggestion above. Let me know what you think. :)

File: src/root-hits.jl (hunk 7 above)
Comment (oschulz) [reply src/root-hits.jl]: `DarioHitsFile` seems fine to me.

**Comment by Luapulu** [comment]
Renaming done.

**Comment by oschulz** [comment]
Good to merge, from your side?

**Comment by Luapulu** [comment]
jup, good to merge :)

**Comment by oschulz** [comment]
Uh, you still have the "test/test.root.hits" file in here - could you take that out? Otherwise it'll blow up the size of the repo forever, once it's merged.

**Comment by Luapulu** [comment]
Done

**Comment by oschulz** [comment]
Nice, here we go ...
